### PR TITLE
feat: limit concurrent downloads

### LINKED ISSUE #39: Download Rate Limiting
Possible to limit download concurrency (e.g. using [p-limit](https://www.npmjs.com/package/p-limit)) as I suspect I'm hitting the rate-limit for npm registry? I need to run several times (as previously downloaded tgz are skipped) before the download ran to completion.

image

Tried monkey-patching the original download https://github.com/Meir017/node-tgz-downloader/blob/9c144bfa7a7acf96cd13ab99e7f7135769f64771/lib/downloader.js#L52-L58

with the following and it worked flawlessly.

```js
  const asyncSequentialDownload = async () => {
    for (let i = 0; i < tarballs.length; i++) {
      const { url, directory } = tarballs[i];
      const position = `${i + 1}/${tarballs.length}`;
      logger(['downloading'.cyan, position], url);
      await _downloadFileWithRetry(url, join(baseDirectory, directory), position, 10);
    }
  }

  return asyncSequentialDownload();
```

### PR DESCRIPTION
**Pull request by deskoh**
Fixes #39.
Wondering whether there should be some default value for download concurrency?

### CODE REVIEW

File: bin/download-tgz [hunk 1]
```diff
@@ -17,11 +17,20 @@ const commands = require('../lib/commands');
 
 program.version(packageJson.version);
 
+const parseConcurrency = (value) => {
```
Comment (Meir017) [inline bin/download-tgz:20]: move this function into `lib`, this file should only describe the API of the CLI tool

File: lib/downloader.js [hunk 2]
```diff
@@ -10,27 +10,28 @@ const urlResolver = require('./url-resolver');
...
 /**
- * @param {{directory: string, registry?: string}} options 
+ * @param {{directory: string, registry?: string}} options
  */
 function downloadFromPackageLock(packageLock, options) {
   const tarballs = [];
   _enumerateDependencies(tarballs, packageLock.dependencies, options);
 
-  return _downloadTarballs(tarballs, options.directory);
+  console.log('options', options);
```
Comment (Meir017) [inline lib/downloader.js:19]: remove this

File: lib/downloader.js [hunk 3]
```diff
@@ -10,27 +10,28 @@ const urlResolver = require('./url-resolver');
...
+  return _downloadTarballs(tarballs, options.directory, options.concurrency);
```
Comment (Meir017) [inline lib/downloader.js:20]: add `concurrency` to the jsdoc in this file

File: bin/download-tgz (hunk 1 above)
Comment (Meir017) [inline bin/download-tgz:20]: we need a default concurrency here, I would suggest `require('os').cpus().length`

**Review by Meir017 — CHANGES_REQUESTED** [review]

File: bin/download-tgz (hunk 1 above)
Comment (deskoh) [reply bin/download-tgz:20]: Do you mean to default the concurrency option value like the following?
```js
.option('-c, --concurrency <concurrency>', 'number of concurrent download', commands.parseConcurrency, cpuCount)
```

Comment (Meir017) [reply bin/download-tgz:20]: yes, something like that

**Review by Meir017 — APPROVED** [review]

**Pull request merged**